File: python/canvastekk_workflow_sdk/workflow/runner.py
"""
Workflow Runner

Local workflow executor that ties together executor, resolver, and level computation.
Accepts a NodeExecutor via constructor (Strategy pattern).
"""

from __future__ import annotations

import asyncio
import time
from enum import StrEnum
from typing import TYPE_CHECKING, Any

from canvastekk_workflow_sdk.context import ExecutionContext
from canvastekk_workflow_sdk.workflow._control_flow import CONTROL_FLOW_HANDLERS
from canvastekk_workflow_sdk.workflow.executor import NodeExecutor
from canvastekk_workflow_sdk.workflow.level import compute_levels
from canvastekk_workflow_sdk.workflow.resolver import resolve_inputs

if TYPE_CHECKING:
    from canvastekk_workflow_sdk.workflow.models import WorkflowSpec


class ErrorPolicy(StrEnum):
    """How the runner handles node execution failures."""

    FAIL_FAST = "fail_fast"
    CONTINUE = "continue"


class NodeResult:
    """Result of executing a single node."""

    __slots__ = ("node_id", "slug", "status", "outputs", "duration_ms", "error", "skipped_reason")

    def __init__(
        self,
        node_id: str,
        slug: str,
        status: str,
        outputs: dict[str, Any] | None = None,
        duration_ms: int = 0,
        error: str | None = None,
        skipped_reason: str | None = None,
    ) -> None:
        self.node_id = node_id
        self.slug = slug
        self.status = status
        self.outputs = outputs
        self.duration_ms = duration_ms
        self.error = error
        self.skipped_reason = skipped_reason


class WorkflowRunResult:
    """Result of executing a complete workflow."""

    __slots__ = ("status", "final_outputs", "node_results", "duration_ms")

    def __init__(
        self,
        status: str,
        final_outputs: dict[str, Any],
        node_results: list[NodeResult],
        duration_ms: int,
    ) -> None:
        self.status = status
        self.final_outputs = final_outputs
        self.node_results = node_results
        self.duration_ms = duration_ms


class WorkflowRunner:
    """Local workflow runner.

    Accepts a ``NodeExecutor`` strategy and runs a workflow spec level-by-level.

    Example::

        executor = InProcessExecutor()
        executor.register("segmentation-v1.0.0", MySegmentNode())
        runner = WorkflowRunner(executor)
        result = runner.run(spec, inputs={"point_cloud": "/data/scan.las"})
    """

    def __init__(
        self,
        executor: NodeExecutor,
        *,
        error_policy: ErrorPolicy = ErrorPolicy.FAIL_FAST,
    ) -> None:
        self._executor = executor
        self._error_policy = error_policy

    def run(
        self,
        spec: WorkflowSpec,
        inputs: dict[str, Any] | None = None,
    ) -> WorkflowRunResult:
        """Execute a workflow synchronously.

        Args:
            spec: The workflow spec to execute.
            inputs: Initial inputs for the START node.

        Returns:
            WorkflowRunResult with final outputs and per-node results.
        """
        return asyncio.run(self.run_async(spec, inputs))

    async def run_async(
        self,
        spec: WorkflowSpec,
        inputs: dict[str, Any] | None = None,
    ) -> WorkflowRunResult:
        """Execute a workflow asynchronously.

        Args:
            spec: The workflow spec to execute.
            inputs: Initial inputs for the START node.

        Returns:
            WorkflowRunResult with final outputs and per-node results.
        """
        start_time = time.perf_counter()
        node_map = {n.id: n for n in spec.nodes}
        node_outputs: dict[str, dict[str, Any]] = {}
        node_results: list[NodeResult] = []
        failed_nodes: set[str] = set()

        if inputs:
            start_nodes = [n for n in spec.nodes if n.slug == "__start__"]
            if start_nodes:
                node_outputs[start_nodes[0].id] = inputs

        levels = compute_levels(spec)

        for level in levels:
            control_ids: list[str] = []
            user_ids: list[str] = []

            for nid in level:
                slug = node_map[nid].slug
                if slug in CONTROL_FLOW_HANDLERS:
                    control_ids.append(nid)
                else:
                    user_ids.append(nid)

            for nid in control_ids:
                node = node_map[nid]
                handler = CONTROL_FLOW_HANDLERS[node.slug]
                resolved = resolve_inputs(nid, spec, node_outputs)
                context = ExecutionContext(run_id="local", node_id=nid)
                t0 = time.perf_counter()
                try:
                    outputs = handler(resolved, context)
                    node_outputs[nid] = outputs
                    node_results.append(
                        NodeResult(
                            node_id=nid,
                            slug=node.slug,
                            status="completed",
                            outputs=outputs,
                            duration_ms=int((time.perf_counter() - t0) * 1000),
                        )
                    )
                except Exception as exc:
                    failed_nodes.add(nid)
                    node_results.append(
                        NodeResult(
                            node_id=nid,
                            slug=node.slug,
                            status="failed",
                            duration_ms=int((time.perf_counter() - t0) * 1000),
                            error=str(exc),
                        )
                    )

            tasks: list[Any] = []
            task_node_ids: list[str] = []

            for nid in user_ids:
                upstream_failed = any(
                    e.from_node in failed_nodes
                    for e in spec.edges
                    if e.to_node == nid
                )
                if upstream_failed or nid in failed_nodes:
                    node = node_map[nid]
                    node_results.append(
                        NodeResult(
                            node_id=nid,
                            slug=node.slug,
                            status="skipped",
                            skipped_reason="upstream_failed",
                        )
                    )
                    failed_nodes.add(nid)
                    continue

                node = node_map[nid]
                if not self._executor.has(node.slug):
                    node_results.append(
                        NodeResult(
                            node_id=nid,
                            slug=node.slug,
                            status="failed",
                            error=f"No executor registered for slug '{node.slug}'",
                        )
                    )
                    failed_nodes.add(nid)
                    if self._error_policy == ErrorPolicy.FAIL_FAST:
                        break
                    continue

                resolved = resolve_inputs(nid, spec, node_outputs)
                context = ExecutionContext(run_id="local", node_id=nid)
                tasks.append(self._executor.execute(node.slug, resolved, context))
                task_node_ids.append(nid)

            if tasks:
                results = await asyncio.gather(*tasks, return_exceptions=True)
                for i, result in enumerate(results):
                    nid = task_node_ids[i]
                    node = node_map[nid]
                    if isinstance(result, Exception):
                        failed_nodes.add(nid)
                        node_results.append(
                            NodeResult(
                                node_id=nid,
                                slug=node.slug,
                                status="failed",
                                error=str(result),
                            )
                        )
                    else:
                        node_outputs[nid] = result
                        node_results.append(
                            NodeResult(
                                node_id=nid,
                                slug=node.slug,
                                status="completed",
                                outputs=result,
                            )
                        )

            if self._error_policy == ErrorPolicy.FAIL_FAST and failed_nodes:
                break

        final_outputs: dict[str, Any] = {}
        for n in spec.nodes:
            if n.slug == "__end__" and n.id in node_outputs:
                final_outputs.update(node_outputs[n.id])

        duration_ms = int((time.perf_counter() - start_time) * 1000)
        status = "completed" if not failed_nodes else "failed"

        return WorkflowRunResult(
            status=status,
            final_outputs=final_outputs,
            node_results=node_results,
            duration_ms=duration_ms,
        )

File: typescript/src/workflow/runner.ts
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import type { WorkflowDefinitionSpec } from "./models.js";
import type { NodeExecutor } from "./executor.js";
import { computeLevels } from "./level.js";
import { resolveInputs } from "./resolver.js";
import { CONTROL_FLOW_HANDLERS } from "./control-flow.js";
import { ExecutionContext } from "../context.js";

export type ErrorPolicy = "fail_fast" | "continue";

export interface NodeResult {
  node_id: string;
  slug: string;
  status: string;
  outputs?: Record<string, unknown> | null;
  duration_ms: number;
  error?: string | null;
  skipped_reason?: string | null;
}

export interface WorkflowRunResult {
  status: string;
  final_outputs: Record<string, unknown>;
  node_results: NodeResult[];
  duration_ms: number;
  output_dir: string | null;
}

/**
 * Executes workflows locally using a node executor.
 */
export class WorkflowRunner {
  private _executor: NodeExecutor;
  private _errorPolicy: ErrorPolicy;
  private _outputDir: string | null;
  private _cleanup: boolean;

  /**
   * Creates a new workflow runner.
   * @param executor - Node executor for running individual nodes
   * @param opts - Runner options
   */
  constructor(
    executor: NodeExecutor,
    opts?: {
      errorPolicy?: ErrorPolicy;
      outputDir?: string;
      cleanup?: boolean;
    },
  ) {
    this._executor = executor;
    this._errorPolicy = opts?.errorPolicy ?? "fail_fast";
    this._outputDir = opts?.outputDir ?? null;
    this._cleanup = opts?.cleanup ?? true;
  }

  /**
   * Executes a workflow specification with the given inputs.
   *
   * Nodes are executed in topological order (levels). The output directory
   * is cleaned up automatically if created by the runner (when `cleanup: true`).
   *
   * @param spec - Workflow specification to execute
   * @param inputs - Initial inputs for the START node
   * @returns Execution result with status, outputs, and node results
   */
  async run(
    spec: WorkflowDefinitionSpec,
    inputs?: Record<string, unknown>,
  ): Promise<WorkflowRunResult> {
    const startTime = performance.now();
    const nodeMap = new Map(spec.nodes.map((n) => [n.id, n]));
    const nodeOutputs: Record<string, Record<string, unknown>> = {};
    const seededOutputs = new Map<string, Record<string, unknown>>();
    const node_results: NodeResult[] = [];
    const failedNodes = new Set<string>();

    let runOutputDir: string;
    let autoCreated = false;

    if (this._outputDir) {
      runOutputDir = this._outputDir;
    } else {
      runOutputDir = mkdtempSync(join(tmpdir(), "wf-runner-"));
      autoCreated = true;
    }

    let status = "failed";
    let final_outputs: Record<string, unknown> = {};
    let duration_ms = 0;
    let result_output_dir: string | null = null;

    try {
      // Seed run inputs for the start node; merged (not clobbered) with the
      // start node's static inputs in the level loop below.
      const startNodes = spec.nodes.filter((n) => n.slug === "__start__");
      if (startNodes.length > 0) {
        seededOutputs.set(startNodes[0].id, { ...(inputs ?? {}) });
      }

      const levels = computeLevels(spec);

      for (const level of levels) {
        const controlIds: string[] = [];
        const userIds: string[] = [];

        for (const nid of level) {
          const slug = nodeMap.get(nid)!.slug;
          if (slug && slug in CONTROL_FLOW_HANDLERS) {
            controlIds.push(nid);
          } else {
            userIds.push(nid);
          }
        }

        for (const nid of controlIds) {
          const node = nodeMap.get(nid)!;
          const slug = node.slug!;
          const handler = CONTROL_FLOW_HANDLERS[slug];
          let resolved: Record<string, unknown>;
          try {
            resolved = resolveInputs(nid, spec, nodeOutputs);
          } catch (exc) {
            failedNodes.add(nid);
            node_results.push({
              node_id: nid,
              slug,
              status: "failed",
              error: `Input resolution failed: ${exc}`,
              duration_ms: 0,
            });
            continue;
          }
          // MERGE semantics: seeded run inputs win over the start node's
          // static inputs — static inputs are preserved and the __start__
          // NodeResult is still recorded.
          if (slug === "__start__" && seededOutputs.has(nid)) {
            resolved = { ...resolved, ...seededOutputs.get(nid) };
          }
          // Per-node subdir prevents same-filename collisions between
          // parallel nodes; absolute-path hand-off via edge outputs preserved.
          const context = new ExecutionContext({ runId: "local", nodeId: nid, outputDir: join(runOutputDir, nid) });
          const t0 = performance.now();
          try {
            const outputs = handler(resolved, context);
            nodeOutputs[nid] = outputs;
            node_results.push({
              node_id: nid,
              slug,
              status: "completed",
              outputs,
              duration_ms: Math.round(performance.now() - t0),
            });
          } catch (exc) {
            failedNodes.add(nid);
            node_results.push({
              node_id: nid,
              slug,
              status: "failed",
              duration_ms: Math.round(performance.now() - t0),
              error: String(exc),
            });
          }
        }

        const tasks: Promise<Record<string, unknown>>[] = [];
        const taskNodeIds: string[] = [];

        for (const nid of userIds) {
          const upstreamFailed = spec.edges.some(
            (e) => e.to_node === nid && failedNodes.has(e.from_node),
          );
          if (upstreamFailed || failedNodes.has(nid)) {
            const node = nodeMap.get(nid)!;
            node_results.push({
              node_id: nid,
              slug: node.slug ?? "",
              status: "skipped",
              skipped_reason: "upstream_failed",
              duration_ms: 0,
            });
            failedNodes.add(nid);
            continue;
          }

          const node = nodeMap.get(nid)!;
          const slug = node.slug;
          if (!slug || !this._executor.has(slug)) {
            node_results.push({
              node_id: nid,
              slug: slug ?? "",
              status: "failed",
              error: slug ? `No executor registered for slug '${slug}'` : "Node has no slug",
              duration_ms: 0,
            });
            failedNodes.add(nid);
            if (this._errorPolicy === "fail_fast") break;
            continue;
          }

          let resolved: Record<string, unknown>;
          try {
            resolved = resolveInputs(nid, spec, nodeOutputs);
          } catch (exc) {
            failedNodes.add(nid);
            node_results.push({
              node_id: nid,
              slug: slug ?? "",
              status: "failed",
              error: `Input resolution failed: ${exc}`,
              duration_ms: 0,
            });
            if (this._errorPolicy === "fail_fast") break;
            continue;
          }
          // Per-node subdir prevents same-filename collisions between
          // parallel nodes; absolute-path hand-off via edge outputs preserved.
          const context = new ExecutionContext({ runId: "local", nodeId: nid, outputDir: join(runOutputDir, nid) });
          tasks.push(this._executor.execute(slug, resolved, context));
          taskNodeIds.push(nid);
        }

        if (tasks.length > 0) {
          const results = await Promise.allSettled(tasks);
          for (let i = 0; i < results.length; i++) {
            const nid = taskNodeIds[i];
            const node = nodeMap.get(nid)!;
            const r = results[i];
            if (r.status === "rejected") {
              failedNodes.add(nid);
              node_results.push({
                node_id: nid,
                slug: node.slug ?? "",
                status: "failed",
                error: String(r.reason),
                duration_ms: 0,
              });
            } else {
              nodeOutputs[nid] = r.value;
              node_results.push({
                node_id: nid,
                slug: node.slug ?? "",
                status: "completed",
                outputs: r.value,
                duration_ms: 0,
              });
            }
          }
        }

        if (this._errorPolicy === "fail_fast" && failedNodes.size > 0) {
          break;
        }
      }

      final_outputs = {};
      for (const n of spec.nodes) {
        if (n.slug === "__end__" && n.id in nodeOutputs) {
          Object.assign(final_outputs, nodeOutputs[n.id]);
        }
      }

      duration_ms = Math.round(performance.now() - startTime);
      status = failedNodes.size === 0 ? "completed" : "failed";
    } finally {
      if (autoCreated && this._cleanup) {
        try {
          rmSync(runOutputDir, { recursive: true, force: true });
        } catch (e) {
          console.warn(`[workflow-runner] Failed to clean up temp dir '${runOutputDir}': ${e}`);
        }
        result_output_dir = null;
      } else {
        result_output_dir = runOutputDir;
      }
    }

    return {
      status,
      final_outputs,
      node_results,
      duration_ms,
      output_dir: result_output_dir,
    };
  }
}
